### ISSUE TITLE
MGCB building Spritefonts wrong on Mac OS X

### ISSUE THREAD

**Issue by kylewlacy** [MonoGame/MonoGame#2108]
After seeing that pull request #2068 added support for building Spritefonts with the newly-introduced `FontTextureProcessor` on Linux and Mac, I decided to try it out. However, after getting MGCB built, and building a test `.spritefont` file and putting it into my game, I noticed a clear issue:
![Spritefont example](https://f.cloud.github.com/assets/1362179/1414565/06a0a3e0-3e93-11e3-9dfe-0a8aaebf165c.png)

I downloaded another pre-compiled Spritefont, and it doesn't exhibit the same issue; meaning that this is an issue with the builder, not the runtime. It appears that the new processor doesn't account for font baselines. However, it may also be important to note that I was also getting the following error with the builder (which is odd, since the `xnb` file was still saved, and could be dropped into the game):

```
Building Font /Library/Fonts/Microsoft Sans Serif.ttf
Building Font /Library/Fonts/Microsoft Sans Serif.ttf
Detected Windows using DXT Compression
System.DllNotFoundException: nvtt
  at (wrapper managed-to-native) Nvidia.TextureTools.Compressor:nvttCreateCompressor ()
  at Nvidia.TextureTools.Compressor..ctor () [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressDxt (Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Boolean generateMipmaps) [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressTexture (Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context, Boolean generateMipmaps, Boolean premultipliedAlpha) [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Processors.FontDescriptionProcessor.Process (Microsoft.Xna.Framework.Content.Pipeline.Graphics.FontDescription input, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context) [0x00000] in <filename unknown>:0 

Build 1 succeeded, 0 failed.

Time elapsed 00:00:02.12.

Unhandled Exception:
System.DllNotFoundException: nvtt
  at (wrapper managed-to-native) Nvidia.TextureTools.Compressor:nvttDestroyCompressor (intptr)
  at Nvidia.TextureTools.Compressor.Finalize () [0x00000] in <filename unknown>:0 
```

I only tested this on Mac OS X (10.9), so I can't say whether or not the same issue exists on Linux or Windows.

**Comment by cra0zy** [comment]
Oh mac has the same nvidia texture tools error as linux... https://github.com/Mono-Game/MonoGame.Dependencies/issues/5

**Comment by SirCmpwn** [comment]
:+1: I get an empty xnb file on Linux, same nvtt error. At some point I was able to get a valid xnb but I can't reproduce it now.

**Comment by cra0zy** [comment]
@SirCmpwn Do you have the newest develop release?

**Comment by SirCmpwn** [comment]
Just compiled it myself a few minutes ago.

**Comment by cra0zy** [comment]
> Just compiled it myself a few minutes ago.

That makes sense now, use the installer, there are few libs that need to be copied that installer copies for you.

Ubuntu: http://teamcity.monogame.net/repository/download/MonoGame_DevelopMac/latest.lastSuccessful/Linux/monogame-sdk.deb?guest=1

Other Linux: http://teamcity.monogame.net/repository/download/MonoGame_DevelopMac/latest.lastSuccessful/Linux/monogame-sdk.run?guest=1

**Comment by SirCmpwn** [comment]
Thanks, I'll try that and report back.

**Comment by SirCmpwn** [comment]
Same error occurs.

**Comment by cra0zy** [comment]
What Linux are you using?

**Comment by SirCmpwn** [comment]
Arch Linux

**Comment by cra0zy** [comment]
What is the exact error you are getting?

**Comment by SirCmpwn** [comment]
```
Build started 2/7/2016 6:16:17 PM

/home/sircmpwn/sources/Universe/DejaVu.spritefont
	Building Font /usr/share/fonts/ttf-dejavu-ib/DejaVuSansCondensed.ttf
	System.DllNotFoundException: nvtt
  at (wrapper managed-to-native) Nvidia.TextureTools.Compressor:nvttCreateCompressor ()
  at Nvidia.TextureTools.Compressor..ctor () <0x41ac0fc0 + 0x0000f> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.DxtBitmapContent.TryCopyFrom (Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent sourceBitmap, Rectangle sourceRegion, Rectangle destinationRegion) <0x41ac0210 + 0x004b3> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent.Copy (Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent sourceBitmap, Rectangle sourceRegion, Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent destinationBitmap, Rectangle destinationRegion) <0x41aaa6e0 + 0x004b3> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent.Copy (Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent sourceBitmap, Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent destinationBitmap) <0x41abce60 + 0x000c3> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent.ConvertBitmapType (System.Type newBitmapType) <0x41abc720 + 0x00253> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.Compress (System.Type targetType, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Boolean generateMipMaps) <0x41abf1c0 + 0x0046f> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressDxt (GraphicsProfile profile, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Boolean generateMipMaps, Boolean sharpAlpha) <0x41abeed0 + 0x000eb> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressTexture (GraphicsProfile profile, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, TextureProcessorOutputFormat format, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context, Boolean generateMipmaps, Boolean sharpAlpha) <0x41abc560 + 0x000bf> in <filename unknown>:0 
  at Microsoft.Xna.Framework.Content.Pipeline.Processors.FontDescriptionProcessor.Process (Microsoft.Xna.Framework.Content.Pipeline.Graphics.FontDescription input, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context) <0x41a97ba0 + 0x00967> in <filename unknown>:0 

```

**Comment by cra0zy** [comment]
What is Arches default lib folder?

Did the script copy nvtt libs? https://github.com/mono/MonoGame/blob/develop/Installers/Linux/RUN/postinstall.sh#L34

**Comment by SirCmpwn** [comment]
Yeah, the libs are there. I've tried putting them in several places:

* The cwd
* /opt/monogame-pipeline
* /usr/lib
* /usr/local/lib

Also made sure that LD_LIBRARY_PATH points to these places when attempting this.

**Comment by SirCmpwn** [comment]
(And the installer puts them in /usr/lib for me (/lib is symlinked here on arch))

**Comment by KonajuGames** [comment]
The NVTT libs were modified to have an rpath of $ORIGIN so that the loader
should look for them in the same directory that MGCB was run from. No
LD_LIBRARY_PATH necessary.

**Comment by SirCmpwn** [comment]
Good to know, but the problem persists.

**Comment by cra0zy** [comment]
> The NVTT libs were modified to have an rpath of $ORIGIN so that the loader
> should look for them in the same directory that MGCB was run from. No
> LD_LIBRARY_PATH necessary.

Linux libs weren't modified for that, original Linux nvtt libs were broken so I recompiled them from source and added them.

**Comment by SirCmpwn** [comment]
Would someone here be willing to compile these spritefonts for me if they have it working? Happy to continue helping identify this problem, but I would also like to make some progress on my project.

https://sr.ht/oAgj.spritefont

https://sr.ht/zhLp.spritefont

**Comment by cra0zy** [comment]
You can compile the fonts by setting the compilation option to Color.

**Comment by SirCmpwn** [comment]
>You can compile the fonts by setting the compilation option to Color.

What? By invoking MGCB manually? How does that work?

**Comment by cra0zy** [comment]
https://github.com/mono/MonoGame/issues/4156#issuecomment-158541108

**Comment by SirCmpwn** [comment]
Thanks!  That works.

**Comment by KonajuGames** [comment]
> recompiled them from source and added them.

The rpath fix was necessary for those who never use the installer. Building from source, the required libs are copied to the output directory.  When you built the libs from source, the rpath was set to your home directory.

https://gist.github.com/KonajuGames/b28b4278af5ed3d858ac

**Comment by cra0zy** [comment]
> https://gist.github.com/KonajuGames/b28b4278af5ed3d858ac

How do you get that info?

**Comment by cra0zy** [comment]
> The rpath fix was necessary for those who never use the installer. Building from source, the required libs are copied to the output directory. When you built the libs from source, the rpath was set to your home directory.
> 
> https://gist.github.com/KonajuGames/b28b4278af5ed3d858ac

I finally understand what was going on, thanks for sharing the info, I won't make the same mistake twice.

**Comment by cra0zy** [comment]
@SirCmpwn I just installed Arch, so I will look into the nvtt bug tomorrow. Today I will do some improvements for the installer(since I noticed icon isn't being installed correctly).

Btw. If you know anything about the proper way for icon installation on Linux, do tell, I couldn't much info on it.

**Comment by cra0zy** [comment]
@SirCmpwn For me nvtt compression works fine on Arch...

I've fixed the icon in #4523 so you can use: http://teamcity.monogame.net/repository/download/MonoGame_DevelopMac/14723:id/Linux/monogame-sdk.run to get the the fixed version.

**Comment by cra0zy** [comment]
Closing this as stuff got fixed up by now.
